### ISSUE TITLE
Variation ", " is not supported by the property type 

### ISSUE THREAD

**Issue by nul800sebastiaan** [umbraco/Umbraco.Cloud.Issues#135]
As described in https://github.com/umbraco/Umbraco.Cloud.Issues/issues/108 people are seeing errors while restoring content: `Variation "Variation "<null>,<null>" is not supported by the property type`.

We need to figure out how to replicate this specific error to be able to figure out what we can do about it.

---
_This item has been added to our backlog [AB#3253](https://umbraco.visualstudio.com/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/3253)_

**Comment by OlePc** [comment]
The error message says `Variation "<null>,<null>"`. Seems like Github is writing `<null>` as blank space.

**Comment by nul800sebastiaan** [comment]
Aha! Good point, thanks! Any ideas on how to reproduce would be great 🙏

**Comment by Prizonic** [comment]
I am haven't tried to setup a new cloud environment to reproduce it but what we did that I think is the problem is the following.

1. When creating the new website we only used english as language.
2. Created some content in that language with allow varying per culture value is false
3. Enabled swedish and set it as default language with no fallback language
4. Created some content
5. Enabled allow varying per culture at all document types.
6. Created some content
7. Getting the error: Variation "Variation "<null>,<null>" is not supported by the property type when we are trying to restore from umbraco cloud dev to our local development environment.

**Comment by clausjensen** [comment]
I'm still unable to reproduce this issue. I've tried a bunch of different things now that I had suspected might be causing it. Also tried follow the exact instructions here + the instructions here with some modifications + instructions from the related issue. I can't make it break doing any of these things.

I'm acknowledging that you have had this issue and would also be very happy to look into it again if it can be somehow consistently reproduced by a workflow you can describe. Until then though, it is a bit hard for me to get any further with debugging and I will mark the issue as closed for now.

This was all tested on new sites running 8.2.0 and Deploy 3.1.5 (both latest at the moment) in Cloud production environment (same as you would be running in as a customer).

**Comment by ronaldbarendse** [comment]
We just had the same problem and it seems this is caused by saving culture variant content/nodes with invariant properties and then changing the property to allow varying by culture. The invariant data is kept in the database (as it should), but when that's the case, it throws this error when deploying.

The following database query returns all invariant property data, while it's property type has variations enabled:
```sql
SELECT * FROM umbracoPropertyData WHERE languageId IS NULL AND segment IS NULL AND propertyTypeId NOT IN (SELECT id FROM cmsPropertyType WHERE variations = 0)
```

After we deleted these rows, the deployment didn't throw the error anymore. So this might be considered a work-around, although you'll obviously lose the data...

**Comment by Zweben** [comment]
> We just had the same problem and it seems this is caused by saving culture variant content/nodes with invariant properties and then changing the property to allow varying by culture. The invariant data is kept in the database (as it should), but when that's the case, it throws this error when deploying.
> 
> The following database query returns all invariant property data, while it's property type has variations enabled:
> 
> ```sql
> SELECT * FROM umbracoPropertyData WHERE languageId IS NULL AND segment IS NULL AND propertyTypeId NOT IN (SELECT id FROM cmsPropertyType WHERE variations = 0)
> ```
> 
> After we deleted these rows, the deployment didn't throw the error anymore. So this might be considered a work-around, although you'll obviously lose the data...

@ronaldbarendse Your conclusion seems to fit with my experience with the same issue. Did you delete the database rows returned by that query on a Live environment or on Dev or local? Do you think if we did this locally or on Dev, we could then deploy the 'fixed' nodes to the Live environment while the issue is still in place on Live? We can't really go in and delete a bunch of data on our Live site.

**Comment by Zweben** [comment]
> 
> 
> > We just had the same problem and it seems this is caused by saving culture variant content/nodes with invariant properties and then changing the property to allow varying by culture. The invariant data is kept in the database (as it should), but when that's the case, it throws this error when deploying.
> > The following database query returns all invariant property data, while it's property type has variations enabled:
> > ```sql
> > SELECT * FROM umbracoPropertyData WHERE languageId IS NULL AND segment IS NULL AND propertyTypeId NOT IN (SELECT id FROM cmsPropertyType WHERE variations = 0)
> > ```
> > 
> > 
> > After we deleted these rows, the deployment didn't throw the error anymore. So this might be considered a work-around, although you'll obviously lose the data...
> 
> @ronaldbarendse Your conclusion seems to fit with my experience with the same issue. Did you delete the database rows returned by that query on a Live environment or on Dev or local? Do you think if we did this locally or on Dev, we could then deploy the 'fixed' nodes to the Live environment while the issue is still in place on Live? We can't really go in and delete a bunch of data on our Live site.

After some careful testing, we ended up having to apply the fix manually on each database, but it worked flawlessly, and everything seems to be working properly with Deploy now.

Thanks again Ronald for sharing this, it was a huge help! #H5YR

**Comment by jmayntzhusen** [comment]
As mentioned above this issue is still occuring, so will reopen so people can try out @ronaldbarendse 's workaround.

Internal issues with further information: 
https://github.com/umbraco/Umbraco-Deploy/issues/410
https://github.com/umbraco/Umbraco-Deploy/issues/406

**Comment by stefankip** [comment]
Just ran into this exact issue and this happened due to the exact reason @ronaldbarendse shared: https://github.com/umbraco/Umbraco.Cloud.Issues/issues/135#issuecomment-576311741

**Comment by lukehook** [comment]
We too are experiencing this issue,

I have a master project that is the baseline for a number of client child projects also and this is currently causing issues as I can't restore to my local project.

Going to try the above query but is there any update on a possible solution for this?

**Comment by FransdeJong** [comment]
I ran into this issue on a non cloud project. I did the query as @ronaldbarendse described. I got one result. Deleted the row and still no joy...

**Comment by Craig-Cooke** [comment]
I also just had this issue running a project in Umbraco cloud.. had to delete 50 rows as suggested higher in the thread to resolve it.

We also have an environment using 2 languages, the content affected was in place before I added the second language, and was a composition type used by most document types.

**Comment by FransdeJong** [comment]
@ronaldbarendse 
If I delete the lines as you described. What am I deleting? 
Active content? Or old content from before the site was Multilingual?

I have 360 lines that I need to delete on live. If it's unused content that isn't cleaned up than it's no issue. Otherwise I need to figure out how I can save the content and solve the issue.

**Comment by ronaldbarendse** [comment]
@FransdeJong Because you've changed the property to store (culture and/or segment) variant content, the previous invariant content isn't used anymore. It's not deleted from the database however, so if you make the property invariant again, you get the 'old' content back (or in a more realistic example: if you accidentally enable variants and disable it again, you don't want the content to be gone).

So if you're not going to use the invariant content anymore, it's safe to delete those rows from the database. I would strongly advise making a backup before doing manual changes in the database, so you can always archive that backup in case you need the content back 😉

**Comment by FransdeJong** [comment]
Thanks for the explanation. It wasn't clear to me what I was deleting.
I'll backup and delete the rows.

Thanks again!

**Comment by alaasaeeddct** [comment]
Hi Dears,
related to this issue still i am trying to fix but no chance to resolve, I have Umbraco 8.6.4 and i can't create any new pages.
could you please support me to resolve this issue.

best regards,

**Comment by alaasaeeddct** [comment]
Hi Dears,
there is no update on this issue it's really serious issue that we need to have a fix for it.

**Comment by nul800sebastiaan** [comment]
Hi @alaasaeeddct unfortunately we haven't been able to get to this one yet, but we're planning on getting this fixed soon. In the mean time I'm afraid you're going to have to use the workaround as described above.

**Comment by alaasaeeddct** [comment]
Hi @nul800sebastiaan I tried this workaround but no luck to fix the issue, this is big issue for me that i am not able to create pages.

kindly, looking for your support at least so I can create new pages.

**Comment by nul800sebastiaan** [comment]
@alaasaeeddct You'll need to provide some details around what you did, what results you got what rows you deleted from the database. I see that @FransdeJong also had issues, maybe he's solved them by now?

**Comment by alaasaeeddct** [comment]
@nul800sebastiaan When i ran the query i got 9000 row in the database, and with this images you can see which rows that i deleted.

![image (1)](https://user-images.githubusercontent.com/71432764/93882072-c410b000-fcf0-11ea-9268-a2dc32bc5732.png)
![image](https://user-images.githubusercontent.com/71432764/93882074-c541dd00-fcf0-11ea-8baf-f86756e551e0.png)
![Screen Shot 2020-08-18 at 1 18 50 PM](https://user-images.githubusercontent.com/71432764/93882075-c541dd00-fcf0-11ea-9bd5-e7d94f142811.png)



![image (2)](https://user-images.githubusercontent.com/71432764/93882061-bfe49280-fcf0-11ea-9ea8-73fce6542619.png)

**Comment by alaasaeeddct** [comment]
@nul800sebastiaan is there any update on this issue, i am stuck with it and can't create any new pages.

**Comment by nul800sebastiaan** [comment]
We don't currently have someone to actively work on this unfortunately. From what I understand is that deleting the property data without a languageID in the DB should be enough to get you going again.

We've got this issue scheduled for next sprint which starts on Monday so hopefully we'll have some more updates for this soon.

**Comment by alaasaeeddct** [comment]
@nul800sebastiaan quick question i am just follow up with error, hope that you start looking into this bug, and i am sure that you will find a permanent solution.

**Comment by liamlaverty** [comment]
Hey,

I've recreated this issue in a non-cloud Umbraco instance, and thought some details may help solve the issue for cloud. Feel free to disregard if it's not helpful. 

Steps to reproduce:

1. Create a document type and include a property with the alias "pageMedia"
2. Set the property's "allow varying by culture" to true.
   - note that in the image below it's set to mandatory, I don't _think_ that's involved in the bug as I've had the same error in non-mandatory instances too
![image](https://user-images.githubusercontent.com/7046713/94993896-05744b80-058c-11eb-9ef0-35a044bed2c8.png)
3. Write the following code

```c# 

string myMainCulture = "en-US"; // set to whatever your main culture is
IContent mySampleDocument = contentService.GetById(mySampleDocumentId); // get a document with that property value
var myMediaFileInUmbraco = Services.MediaService.GetById(myMediaId); // get the media file to be saved to the document
var myPageMediaUdi =  Udi.Create(Constants.UdiEntityType.Media, myMediaFileInUmbraco.Key); // get the UDI for that media item

// -- -- -- relevant stuff starts here: -- -- -- //

// set the property on the document, the error in the code is here
mySampleDocument.SetValue("pageMedia", myPageMediaUdi);

// do (probably lots of) other stuff

// save the content to Umbraco, the error is thrown by Umbraco here
Services.ContentService.SaveAndPublish(mySampleDocument, myMainCulture );
```


The SaveAndPublish method is the one returning the `"Variation "<null>,<null>"`. It's happening because SetValue should have a culture on it. You can pass in either `myMainCulture` or `"*"` to set to all. 

To stop the error from being thrown, the following code will work: 

```c# 
// set the property on the document, and assign a culture to it
mySampleDocument.SetValue("pageMedia", myPageMediaUdi, myMainCulture);

// save the content to Umbraco, the error is no longer thrown here
Services.ContentService.SaveAndPublish(mySampleDocument, myMainCulture );
```

For what it's worth, I think Umbraco is behaving correctly here - it's not a bug - it's just that the error message isn't very helpful in this specific scenario. It'd be neat if it said something along the lines of "language variation on property with alias `pageMedia` is unset, please set a culture". Unfortunately you can't just set the culture on every `SetValue()` or `SaveAndPublish()` call, because you get the opposite error on properties without "allow varying by culture" set.

Hope that helps either with this issue, or if someone else finds this page through Google/Bing's search results.

**Comment by ronaldbarendse** [comment]
The error message `Variation "<null>,<null>" is not supported by the property type` is quite straightforward: the property doesn't support setting invariant values, but if you do, `SetValue` throws this exception: https://github.com/umbraco/Umbraco-CMS/blob/8cba80913a46628778b477b39d1e87bb4b5e1fca/src/Umbraco.Core/Models/Property.cs#L291-L310

Like @liamlaverty notes, this is not the real issue: it's Umbraco Deploy trying to set a value using a non-valid variation. If the content types differ between environments, you would get a schema mismatch, so this must be something with how the content is stored in the artifact (e.g. not looking at the content/property variance, but only how it's currently stored).

**Comment by alaasaeeddct** [comment]
Morning Every one, this bug or issue really should be fixed as it's became a big error and makes difficulties using Umbraco.
hope that we can find it.

**Comment by warrenbuckley** [comment]
Hello all in this thread 👋 
Based on various peoples notes, I am unable to reproduce this issue with Cloud using the following versions:

**CMS: 8.8.0
Deploy: 3.5.1**

Can you all please confirm to me what version of CMS & Deploy you are using, so I can see if this may have been fixed?

If anyone is able to give very precise steps for me to follow or retry it would be appreciated, but this is what I have done, but please let me know if  I am missing a step or details that cause this problem for you.

- Cloud site Dev & Live
- Create doctype
- Add one textstring prop
- Add a new language (DK)
- Mark doctype as allow variants
- Add new prop (media picker) that does vary
- Deploy UDA schema
- Deploy content to Live - to ensure everything sync before we begin
- Mark textstring prop as allow variants
- Re-deploy UDA schema
- Update content on Dev for both languages
- Save & deploy to Live
- All ok

**Comment by FransdeJong** [comment]
I got into this situation by doing the following:

- Start new cloud project
- Create doctypes without variations
- Push to live
- Create content on live 
- Save and Publish on live
- Change doctype to allow variations
- Change properties to allow variations
- Push to live
- Pull content from live to stage

I don't know the exact version but it was probably 8.6.4 and the latest deploy.

If I can help to try and reproduce please let me know.

**Comment by ronaldbarendse** [comment]
@warrenbuckley You need to have invariant data stored for a property and then make it vary by culture, so the steps from @FransdeJong should indeed reproduce the error.

My guess is that the content artifact is created based on the stored data, instead of its current configuration. This would result in Umbraco Deploy trying to restore invariant data into a culture variant property, hence the exception message returned by `SetValue()`.

**Comment by warrenbuckley** [comment]
@FransdeJong have you been able to reproduce this lately with the latest on Umbraco Cloud & a new project which is 8.8.0. This is what I have done without any repro issues.

As I need some verification to know if this has been fixed for you if you can not reproduce the same problem in 8.8.0 as opposed to 8.6.5, please.

**Comment by FransdeJong** [comment]
@warrenbuckley I haven't tried it there and I don't have anything running above 8.6.5 at the moment.
I could set up a project and test it against 8.8 but I don't expect different results if you did the same as I did...

**Comment by warrenbuckley** [comment]
@FransdeJong I have created this and tried to follow along in a new Umbraco 8.8 site on cloud. See the attached video.

[Link to Video](https://umbracocloud-my.sharepoint.com/:v:/g/personal/warren_umbraco_dk/EYPRuH_IPLBOs6kO4vNO7QABSOa3M8qNz-WF2TmHOrNERw?e=MloBxj)
**Video File Password** = `gh-135`

I have created a blank cloud project called **Testing with Frans** that I have invited you to, with your webwonders.nl email address.

If you could please try to set up & reproduce this in this blank/empty project for me then that would be fantastic, just in case I am missing or misunderstanding one or more of your steps to reproduce this.

**Comment by FransdeJong** [comment]
@warrenbuckley

I'm looking at it right now.
I see I forgot to mention adding the language as you said in the video so maybe that's the issue.

I'll try to reproduce on the project and let you know.

**Comment by FransdeJong** [comment]
Error still exists in V8.8

These are the exact steps I did to reproduce on the new project:
- Dev: Create compositions
- Dev:  Create doctypes
- Push changes to Live
- Transfer content to live
- Dev: Add language
- Dev: Set doctypes to allow variants
- Dev: Set properties to allow variants
- Push changes to live
- Full restore content from live to stage

![image](https://user-images.githubusercontent.com/29602162/95208218-166dc880-07e9-11eb-8556-bee56728b0ca.png)

**Comment by warrenbuckley** [comment]
Ok thanks for helping to repro it for us. I can also use this project to help us debug the problem as well :)

**Comment by FransdeJong** [comment]
No problem, let's hope this issue can be solved before its anniversary 😈😉

**Comment by warrenbuckley** [comment]
No not currently, we can repro with that test cloud project, if we need anything I'll be in touch via this thread :)

**Comment by clausjensen** [comment]
@FransdeJong Hi Frans. Sorry to bother you again about this, but I'm trying to get to the bottom of this problem and need a bit more help.

While I can "reproduce" the error by using the project Warren set up, I would really like to be able to reproduce from scratch in a completely fresh project, with instructions on how to do so that I can follow to get to the point where it fails.

Doing this helps me isolate the problem as much as possible and also helps a lot in the process of debugging the cause of the error. It is also very very helpful to be able to get to the point where it fails, with as few "moving parts" as possible - meaning the Starterkit in the project Warren set up is not really the best fit for this as it has a million different parts.

I've tried a bunch of times following your step by step instructions here and I simply cannot make it fail, when doing that. I've also tried following the previously posted instructions and I've even tried adding in steps that I thought would make it fail, with no luck.

Would it be possible for you to once again (and I'm sorry to have you do it all over again), give complete step by step instructions on exactly what I need to do on a blank project (not installing Starterkit) to get it to a point where I could hopefully have just one document of one document type with one property, not being able to be deployed?

**Comment by FransdeJong** [comment]
@clausjensen Can you add me to a blank cloud install? I'll do the same as I did with the starter kit and report back a reliable reproduction scenario.
We can also plan a call and walk through it together?

**Comment by clausjensen** [comment]
@FransdeJong of course!!

I have added you to a project now and you should be getting an email to confirm. I think the key thing here is to just have every single step (even if you end up doing something and then doing something else, just note it down) listed.... and please.. just dumb it down for me 😃 I'd rather have too much info than too little.

I'm a little busy today to do a call, but if it doesn't make sense with just written instructions we can set up something later this week.

**Comment by FransdeJong** [comment]
@clausjensen I'll take a look later today and report back here.

**Comment by clausjensen** [comment]
@FransdeJong Thanks - your help is much appreciated 😃

**Comment by FransdeJong** [comment]
I tried the following steps to reproduce.

- Dev: Create compositions (PageContent)
- Dev:  Create doctypes (Home)
- Dev: Add Composition to doctype (PageContent to Home)
- Dev: Create Content node using the doctype (Home)
- Dev: Create content in the node (Home), Save and Publish
- Push changes to Live 
- Transfer content to live
- Dev: Add language
- Dev: Set doctypes to allow variants
- Dev: Set properties to allow variants
- Push changes to live
- Full restore content from live to stage 

These steps didn't reproduce the error.

My thought is:
- The issue is possibly resolved in V8.8 when storing new data
- The starterkit (and possibly other packages) stores the data in the old way.
- Deploy should be updated to handle the old way of storing data when it encounters it because of the data can be stored in the old way?

The same will happen with older installs of Umbraco that have to go multilingual after containing existing content.
So this issue should be handled in Deploy as well in my opinion.

**Comment by alaasaeeddct** [comment]
Dears
is this issue has been resolved, it's really hard to work on the website with this error, is there any way to solve it.
really need support to solve this.

**Comment by FransdeJong** [comment]
Not to my knowledge.

**Comment by clausjensen** [comment]
@alaasaeeddct Not yet - we're still working on it!

**Comment by ronaldbarendse** [comment]
> is there any way to solve it

See the above comments, especially https://github.com/umbraco/Umbraco.Cloud.Issues/issues/135#issuecomment-576311741.

TLDR: You need to remove property data that isn't used anymore, e.g. invariant data for a property that's made culture variant (or the other way around). Umbraco keeps this data, as you might want to revert back to the previous configuration (invariant/vary by culture) without losing any data.

**Comment by galaxxy22** [comment]
> We just had the same problem and it seems this is caused by saving culture variant content/nodes with invariant properties and then changing the property to allow varying by culture. The invariant data is kept in the database (as it should), but when that's the case, it throws this error when deploying.
> 
> The following database query returns all invariant property data, while it's property type has variations enabled:
> 
> ```sql
> SELECT * FROM umbracoPropertyData WHERE languageId IS NULL AND segment IS NULL AND propertyTypeId NOT IN (SELECT id FROM cmsPropertyType WHERE variations = 0)
> ```
> 
> After we deleted these rows, the deployment didn't throw the error anymore. So this might be considered a work-around, although you'll obviously lose the data...

This also fits my experience with having created some content-nodes and later having made the doc-types culture variant. Deleted the entries from the query locally and next deployed succesfully to the live environment.

**Comment by alaasaeeddct** [comment]
Hi dears,
is there anything we can do to help in solving this issue.

best regards,

**Comment by alaasaeeddct** [comment]
Dears,
Is there any chance to solve this.

Best regards,

**Comment by alaasaeeddct** [comment]
Dear @nul800sebastiaan 
hope that you are doing well, and please we are looking for the solution for this issue that solve our issues, many users have same issue and we are stuck really as we can't create any new pages, blogs and etc...

best regards

**Comment by nikolajlauridsen** [comment]
We're still looking into this issue, but since we can't reproduce the error, it's very hard to find the root cause, unfortunately. 

I've tried creating a build with additional logging around the code where the error stems from, it would be incredibly helpful if you could try and set logging to "Debug" with this build and post the logs for when the error occurs. 

To use the build you need to update the DLLs in your development cloud sites bin folder with the ones in the deploy/bin folder, and transfer the changes to live. 

[UmbracoDeploy.v3.5.3-alpha-logging.zip](https://github.com/umbraco/Umbraco.Cloud.Issues/files/5855772/UmbracoDeploy.v3.5.3-alpha-logging.zip)

**Comment by nul800sebastiaan** [comment]
I've tried a few scenarios as well but can't reproduce either, I would be very curious to look at a Cloud site experiencing this behavior, feel free to add a comment here to let us know if anyone still sees this error! A possibly related fix in the order of restoring items is coming in Deploy 3.5.4 in 3 days.

I will close this issue for the time being, but if it can still be reproduced with Deploy 3.5.4 then we're happy to have another look.

**Comment by MichaelNielsenDK** [comment]
We're having this issue on a solution. 

Umbraco 8.11.1
Deploy 3.5.4

Curiously, if you via Partial Restore, pick the Umbraco node mentioned in the error, it restores that single node without errors. 🤔🤷‍♂️ 

If you need access to the Cloud project, then just contact me.
